# 0.1 Importing libraries

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 0.2 Importing data files

In [2]:
#Setting path
path = r'C:\Users\skhra\DataAnalytics\Feb2025_InstacartBasketAnalysis'

In [4]:
#Importing orders.csv 
df_ord = pd.read_csv(os.path.join(path, '02- Data', 'OriginalData', 'orders.csv'), index_col = False)

In [5]:
df_ord.shape

(3421083, 7)

In [6]:
#Importing products.csv
df_prod = pd.read_csv(os.path.join(path, '02- Data', 'OriginalData', 'products.csv'), index_col = False)

In [7]:
df_prod.shape

(49693, 5)

In [6]:
#Dropping unnecessary column
df_ord = df_ord.drop(columns = ['eval_set'])

In [7]:
#Counting values
df_ord['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [8]:
#Renaming a column in orders
df_ord.rename(columns = {'order_dow':'order_day_of_week'},inplace = True)

In [9]:
df_ord.head(5)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [10]:
#Importing departments.csv
df_dep = pd.read_csv(os.path.join(path, '02- Data', 'OriginalData', 'departments.csv'), index_col = False)

In [11]:
df_dep.head(5)

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

# 0.3 Wrangling department data

In [12]:
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [13]:
df_dep_t = df_dep.T

In [14]:
df_dep_t.head(3)

0
department_id  department
1                  frozen
2                   other

In [15]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [16]:
new_header = df_dep_t.iloc[0]

In [17]:
new_header

0    department
Name: department_id, dtype: object

In [18]:
df_dep_t_new = df_dep_t[1:]

In [19]:
df_dep_t_new.head(5)

0
1   frozen
2    other
3   bakery
4  produce
5  alcohol

In [20]:
df_dep_t_new.columns=new_header

In [21]:
df_dep_t_new.head(5)

department_id department
1                 frozen
2                  other
3                 bakery
4                produce
5                alcohol

In [22]:
data_dict = df_dep_t_new.to_dict('index')

In [23]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [24]:
df_prod.head(5)

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [25]:
df_snacks = df_prod[df_prod['department_id']==19]

In [26]:
df_snacks.head(5)

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

# 0.4 Wrangling orders dataset

In [27]:
df_ord.head(5)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [28]:
df_ord.dtypes

order_id                    int64
user_id                     int64
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [29]:
#Chnaging data types of order_id and user_id
df_ord['order_id']=df_ord['order_id'].astype('str')
df_ord['user_id']=df_ord['user_id'].astype('str')

In [30]:
df_ord.dtypes

order_id                   object
user_id                    object
order_number                int64
order_day_of_week           int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [31]:
#Renaiming one of the columns
df_ord.rename(columns = {'days_since_prior_order' : 'days_from_previous_order'}, inplace = True)

In [32]:
df_ord.head(4)

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0  2539329       1             1                  2                  8   
1  2398795       1             2                  3                  7   
2   473747       1             3                  3                 12   
3  2254736       1             4                  4                  7   

   days_from_previous_order  
0                       NaN  
1                      15.0  
2                      21.0  
3                      29.0

In [33]:
#Counting orders by hour: It turns out that most of orders are made at 10  
df_ord['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

# 0.5 Data dictionary

In [34]:
#Creating data dictionary
data_dict = df_dep_t_new.to_dict('index')

In [35]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [36]:
#Breakfast item sales
df_breakfast = df_prod[df_prod['department_id']==14]

In [37]:
df_breakfast.head(5)

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

In [38]:
#Items for dinner parties (departments alcoho = 5, deli=20, beverages=7 and meat/seafood=12)
df_dinner = df_prod[df_prod['department_id'].isin([5,7,12,20])]

In [39]:
df_dinner

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

This dataframe df_dinner has 7650 rows in total

In [40]:
#What about customer with user_id =1?
df_customer1 = df_ord[df_ord['user_id']=='1']

In [41]:
df_customer1

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329       1             1                  2                  8   
1   2398795       1             2                  3                  7   
2    473747       1             3                  3                 12   
3   2254736       1             4                  4                  7   
4    431534       1             5                  4                 15   
5   3367565       1             6                  2                  7   
6    550135       1             7                  1                  9   
7   3108588       1             8                  1                 14   
8   2295261       1             9                  1                 16   
9   2550362       1            10                  4                  8   
10  1187899       1            11                  4                  8   

    days_from_previous_order  
0                        NaN  
1                       15.0  
2                       21.0  
3                       29.0  
4                       28.0  
5                       19.0  
6                       20.0  
7                       14.0  
8                        0.0  
9                       30.0  
10                      14.0

In [42]:
#Some statistics on customer1
df_customer1.describe()

order_number  order_day_of_week  order_hour_of_day  \
count     11.000000          11.000000          11.000000   
mean       6.000000           2.636364          10.090909   
std        3.316625           1.286291           3.477198   
min        1.000000           1.000000           7.000000   
25%        3.500000           1.500000           7.500000   
50%        6.000000           3.000000           8.000000   
75%        8.500000           4.000000          13.000000   
max       11.000000           4.000000          16.000000   

       days_from_previous_order  
count                 10.000000  
mean                  19.000000  
std                    9.030811  
min                    0.000000  
25%                   14.250000  
50%                   19.500000  
75%                   26.250000  
max                   30.000000

Very basic conclusion about customer1 behavior is that he/she orders only within days 1 -4 of the weeks in the time interval between 7 and 16. Mean hour about 10 means most of the orders correspond to mornings. 

# 0.6 Export modified data sets

In [43]:
#Exploring data sets
df_ord.to_csv(os.path.join(path, '02- Data', 'PreparedData', 'orders_wrangled.csv'))
df_dep_t_new.to_csv(os.path.join(path, '02- Data', 'PreparedData', 'departments_wrangled.csv'))              

In [55]:
df_ord1 = pd.read_csv(os.path.join(path, '02- Data', 'PreparedData', 'orders_wrangled.csv'), index_col=0)

In [56]:
df_ord1.head(5)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_from_previous_order  
0                       NaN  
1                      15.0  
2                      21.0  
3                      29.0  
4                      28.0